File: implementations/fortran/string-search/v8-knuth-morris-pratt-d7l8/kmp.f90
program kmp_test
  implicit none
  call expect('ababcabcabababd', 'ababd', [10])
  call expect('aaaaa', 'aa', [0, 1, 2, 3])
  call expect_empty('abcdef', 'gh')
  call expect('abc', 'abc', [0])
  print *, 'fortran kmp ok'
contains
  subroutine kmp_search(text, pattern, matches, count)
    character(len=*), intent(in) :: text, pattern
    integer, intent(out) :: matches(:)
    integer, intent(out) :: count
    integer :: lps(len(pattern)), i, length, ti, pj
    if (len(pattern) == 0) then
      count = 0
      return
    end if
    lps = 0
    length = 0
    i = 2
    do while (i <= len(pattern))
      if (pattern(i:i) == pattern(length + 1:length + 1)) then
        length = length + 1
        lps(i) = length
        i = i + 1
      else if (length /= 0) then
        length = lps(length)
      else
        lps(i) = 0
        i = i + 1
      end if
    end do
    count = 0
    ti = 1
    pj = 1
    do while (ti <= len(text))
      if (text(ti:ti) == pattern(pj:pj)) then
        ti = ti + 1
        pj = pj + 1
        if (pj > len(pattern)) then
          count = count + 1
          matches(count) = ti - pj
          pj = lps(pj - 1) + 1
        end if
      else if (pj /= 1) then
        pj = lps(pj - 1) + 1
      else
        ti = ti + 1
      end if
    end do
  end subroutine kmp_search

  subroutine expect(text, pattern, expected)
    character(len=*), intent(in) :: text, pattern
    integer, intent(in) :: expected(:)
    integer :: matches(16), count
    call kmp_search(text, pattern, matches, count)
    if (count /= size(expected)) error stop 'count mismatch'
    if (count > 0) then
      if (any(matches(1:count) /= expected)) error stop 'match mismatch'
    end if
  end subroutine expect

  subroutine expect_empty(text, pattern)
    character(len=*), intent(in) :: text, pattern
    integer :: matches(16), count
    call kmp_search(text, pattern, matches, count)
    if (count /= 0) error stop 'empty mismatch'
  end subroutine expect_empty
end program kmp_test

File: implementations/rust/string-search/v8-knuth-morris-pratt-d7l8/kmp.rs
fn kmp_search(text: &str, pattern: &str) -> Vec<usize> {
    if pattern.is_empty() {
        return Vec::new();
    }
    let t = text.as_bytes();
    let p = pattern.as_bytes();
    let mut lps = vec![0usize; p.len()];
    let mut len = 0usize;
    let mut i = 1usize;
    while i < p.len() {
        if p[i] == p[len] {
            len += 1;
            lps[i] = len;
            i += 1;
        } else if len != 0 {
            len = lps[len - 1];
        } else {
            lps[i] = 0;
            i += 1;
        }
    }
    let mut matches = Vec::new();
    let mut ti = 0usize;
    let mut pj = 0usize;
    while ti < t.len() {
        if t[ti] == p[pj] {
            ti += 1;
            pj += 1;
            if pj == p.len() {
                matches.push(ti - pj);
                pj = lps[pj - 1];
            }
        } else if pj != 0 {
            pj = lps[pj - 1];
        } else {
            ti += 1;
        }
    }
    matches
}

fn main() {
    assert_eq!(kmp_search("ababcabcabababd", "ababd"), vec![10]);
    assert_eq!(kmp_search("aaaaa", "aa"), vec![0, 1, 2, 3]);
    assert_eq!(kmp_search("abcdef", "gh"), Vec::<usize>::new());
    assert_eq!(kmp_search("abc", "abc"), vec![0]);
    println!("rust kmp ok");
}
